[BOUNTY] Implements atom integrity for walls, (which is needed to do) Upgraded Emitters can now tear down walls! Take Two!

Singul0 opened the pull request:
## About The Pull Request
Ports yogstation13/Yogstation#21336
Retake of #5706
This PR is a double bounty, makes it so that walls behaves the same way as any other structures or machineries, being able to be damaged and destroyed by normal weapons instead of very specific ways/deconstruction

Normal walls have 300 damage + armor, with a damage minimum for it to take effect of 25
Normal walls have 600 damage + armor, with a damage minimum for it to take effect of 75

Emitters with upgraded lasers can now be able to tear down walls, stock parts emitters still functions the same. Below are the damage table needed to tear a wall down

T1 laser: cannot damage walls, operates normally.
T2: 20 hits to break normal walls, r wall or stronger cannot be damaged.
T3: 10 hits to kill normal walls, r wall or stronger cannot be damaged.
T4: 5 hits to kill normal walls, r walls or strong take 10 hits to be killed.
T5: 3 hits to kill normal, 7 to kill r walls or higher
Emagging the emitters increases the parts tier above by one (which is how you'd get T5) at the cost of 5% damage to the emitters. This functionality only applies if the projectile emitted by the emitters is the default hitscan emitters shots.

Pooba said that for my first PR to go through, we'd need to implement atom integrity, which they'd gracefully posted a bounty for.
## Why It's Good For The Game
Here is it in action:
https://github.com/user-attachments/assets/34b999b0-a7f6-4fa7-ac26-5e804e3ba5ad

Walls being able to use the integrity system means destroying them in a way that isn't instantaneous  and no longer needs to rely on tons of snowflake code for each instance/case of something breaking a wall. It also means they aren't totally invincible to damage, which I think could allow for some fun interactions.

Anyhow, as for the emitter PR. i'll just copy this verbatim: You know how the emitters looks like a heavy, industrial laser cannon that you'd be able to see be used as an impromptu blaster? How come can it not damage walls? Anyways, this adds a new way of breaching walls besides thermite, tools or c-4/x-4, Albeit a pretty slow and loud one at that. More tools for engineer's arsenal is good!
## Changelog
:cl:
add: Walls now have atom integrity, being able to be attacked and destroyed like other machineries and structures. they're a tough nut to crack though
 add: Emitters can now tear down walls! (once upgraded)
/:cl:

LikeLakers2 reviewed (changes requested):
Please modularize your changes, or use `//MONKESTATION EDIT` commentns where you can't.

LikeLakers2 commented on code/game/turfs/closed/wall/reinf_walls.dm:
```diff
@@ -8,11 +8,12 @@
 	density = TRUE
 	turf_flags = IS_SOLID
 	smoothing_flags = SMOOTH_BITMASK
-	hardness = 10
```
```
	//hardness = 10 //MONKESTATION REMOVAL: Walls now use atom integrity
```

Singul0 commented:
> Please modularize your changes, or use `//MONKESTATION EDIT` comments where you can't/don't know how to.

Done, unfortunately. Due to this being a refactor and all, can't really seperate it into seperate files

LikeLakers2 commented on code/game/turfs/closed/walls.dm:
```diff
@@ -108,10 +126,6 @@
 	if(SSstation_coloring.wall_trims)
 		trim_color = SSstation_coloring.get_default_color()
 
-/turf/closed/wall/atom_destruction(damage_flag)
-	. = ..()
-	dismantle_wall(TRUE, FALSE)
-
```
```
/* //MONKESTATION REMOVAL: Walls now use atom integrity
/turf/closed/wall/atom_destruction(damage_flag)
	. = ..()
	dismantle_wall(TRUE, FALSE)
*/
```

Singul0 replied:
ehh, there is a code that edits this further down, I just moved it closer to edit lines for easier clarity do I still add this

LikeLakers2 replied:
Ah, if that's the case, then don't worry - I didn't look too closely.

LikeLakers2 reviewed (changes requested):
Please also mark your removals, for clarity - I've included a couple examples of what I mean.

LikeLakers2 commented:
> Due to this being a refactor and all, can't really seperate it into seperate files

Remember that you can redefine procs - so if you wanted to comment out the tgstation code, and put the new code under `monkestation/`, that would work too.

Singul0 commented:
> > Due to this being a refactor and all, can't really seperate it into seperate files
> 
> Remember that you can redefine procs in DM - so if you wanted to comment out the tgstation code, and put the new code under `monkestation/`, that would work too.
> 
> Again, sorry for the hassle.

It's fine, don't worry about it. I added docs for the proc that I deleted, I think the rest are just procs that I edited and moved to group edits closer

LikeLakers2 commented on code/game/turfs/closed/wall/reinf_walls.dm:
```diff
@@ -8,6 +8,7 @@
 	density = TRUE
 	turf_flags = IS_SOLID
 	smoothing_flags = SMOOTH_BITMASK
+	hardness = 10
```
Noticed you re-added this... was it intentional? Or did you mean to comment it out?

Singul0 replied:
Ah yeah, that removal was intentional. I wanted to tie it to the refactor but figured that it should be done later

LikeLakers2 reviewed (approved):
Thank you! :) If I spot anything else that needs changing, I'll make a new comment - but I don't see anything right now.

Shoddd commented:
not to major but can only deconstruct in combat mode, feel like it would be vice versa where you should hit if in combat mode, deconstruct if not

Singul0 commented:
> not to major but can only deconstruct in combat mode, feel like it would be vice versa where you should hit if in combat mode, deconstruct if not

It deconstructs in normal mode now, though it's a bit redundant since welding tools doesnt have high enough damage to do anything to the walls anyways

Absolucy commented:
does hitting walls with items respect `demolition_mod`?

Singul0 commented:
> does hitting walls with items respect `demolition_mod`?

whoops fixed now
